cpp/oneapi/dal/test/engine/dataframe_common.hpp

```diff
@@ -0,0 +1,241 @@
+    void add_field(const std::string& name, T&& value) {
+        user_fields_.insert_or_assign(name, std::any{ std::forward<T>(value) });
+    }
+
+    void remove_field(const std::string& name) {
+        user_fields_.erase(name);
+    }
+
+    template <typename T>
+    std::optional<T> get_field(const std::string& name) const {
+        const auto it = user_fields_.find(name);
+        if (it == user_fields_.end()) {
+            return std::nullopt;
+        }
+
+        try {
+            return std::any_cast<T>(it->second);
+        }
+        catch (const std::bad_any_cast&) {
+            return std::nullopt;
+        }
+    }
+
+private:
+    array<float> array_;
+    std::int64_t row_count_;
+    std::int64_t column_count_;
+    std::unordered_map<std::string, std::any> user_fields_;
+};
+
+class dataframe {
+public:
+    explicit dataframe(const array<float>& data, std::int64_t row_count, std::int64_t column_count)
+            : dataframe(new dataframe_impl{ data, row_count, column_count }) {}
+
+    explicit dataframe(dataframe_impl* impl) : impl_(impl) {}
+
+    template <typename Float>
+    table get_table(const std::string& table_type) const;
+
+    template <typename Float>
+    table get_table(host_test_policy& policy, const std::string& table_type) const {
+        return get_table<Float>(table_type);
+    }
+
+#ifdef ONEDAL_DATA_PARALLEL
+    template <typename Float>
+    table get_table(device_test_policy& policy, const std::string& table_type) const;
+#endif
+
+    std::int64_t get_row_count() const {
+        return impl_->get_row_count();
+    }
+
+    std::int64_t get_column_count() const {
+        return impl_->get_column_count();
+    }
+
+    std::size_t get_count() const {
+        return impl_->get_count();
+    }
+
+    std::size_t get_size() const {
+        return impl_->get_size();
+    }
+
+    const array<float>& get_array() const {
+        return impl_->get_array();
+    }
+
+    template <typename T>
+    void add_field(const std::string& name, T&& value) const {
+        impl_->add_field(name, std::forward<T>(value));
+    }
+
+    void remove_field(const std::string& name) const {
+        impl_->remove_field(name);
+    }
+
+    template <typename T>
+    std::optional<T> get_field(const std::string& name) const {
+        return impl_->template get_field<T>(name);
+    }
+
+    template <typename T, typename Op>
+    T get_or_add_field(const std::string& name, Op&& op) const {
+        const auto optional_value = impl_->template get_field<T>(name);
+        if (optional_value) {
+            return optional_value.value();
+        }
+        else {
+            const auto value = op();
+            impl_->add_field(name, value);
+            return value;
+        }
+    }
+
+private:
+    mutable dal::detail::pimpl<dataframe_impl> impl_;
+};
+
+class dataframe_builder_action {
+public:
+    virtual ~dataframe_builder_action() = default;
+    virtual std::string get_opcode() const = 0;
```

Comment:
Is opcode just a serialized form of operation?

Reply:
This is some identifier of `dataframe_builder_action` that is introduced for efficient caching and dataframe reuse across tests. A set of `dataframe_builder_action` builds a sequence of actions to be performed by builder (virtual processor), opcode is very popular term in this context. https://en.wikipedia.org/wiki/Opcode